zkvm/src/txlog.rs

```diff
@@ -76,6 +98,84 @@ impl TxID {
         entry.commit_to_transcript(&mut t);
         t.challenge_bytes(b"merkle.leaf", result);
     }
+
+    /// Builds a precomputed Merkle tree for the entries.
+    pub fn build_tree(list: &[Entry]) -> (Self, MerkleRoot) {
+        let t = Transcript::new(b"ZkVM.txid");
+        let root = Self::build_node(t, list);
+        (
+            TxID(root.node),
+            MerkleRoot {
+                root,
+                size: list.len(),
+            },
+        )
+    }
+
+    fn build_node(mut t: Transcript, list: &[Entry]) -> MerkleNode {
+        match list.len() {
+            0 => {
+                let mut leaf = [0u8; 32];
+                Self::empty(t, &mut leaf);
+                return MerkleNode {
+                    node: leaf,
+                    left: None,
+                    right: None,
+                };
+            }
+            1 => {
+                let mut leaf = [0u8; 32];
+                Self::leaf(t, &list[0], &mut leaf);
+                return MerkleNode {
+                    node: leaf,
+                    left: None,
+                    right: None,
+                };
+            }
+            n => {
+                let k = n.next_power_of_two() / 2;
+                let mut node = [0u8; 32];
+                let left = Self::build_node(t.clone(), &list[..k]);
+                let right = Self::build_node(t.clone(), &list[k..]);
+                t.commit_bytes(b"L", &left.node);
+                t.commit_bytes(b"R", &right.node);
+                t.challenge_bytes(b"merkle.node", &mut node);
+                return MerkleNode {
+                    node: node,
+                    left: Some(Box::new(left)),
+                    right: Some(Box::new(right)),
+                };
+            }
+        }
+    }
+
+    /// Verifies that an entry satisfies the Merkle proof of inclusion
+    /// for a given TxID
+    pub fn verify_proof(&self, entry: Entry, proof: Vec<MerkleHash>) -> Result<(), VMError> {
+        let transcript = Transcript::new(b"ZkVM.txid");
+        let mut result = [0u8; 32];
+        Self::node(transcript.clone(), &[entry], &mut result);
```

Comment:
Should it be a leaf instead of an intermediary node?
```suggestion
        Self::leaf(transcript.clone(), &[entry], &mut result);
```